### PR TITLE
bugfix: clear all cached object when got `delete` action, and removing stale boolean data in table `self.values` with a safer way

### PR DESCRIPTION
**Pull request by membphis**
### What this PR does / why we need it:

If the administrator user creates and deletes the same route multiple times, the new route will not be correctly synchronized to the APISIX gateway node, resulting in some request route matching failures and a 404 response.

bugfix: clear all cached data when get delete action, and removing stale boolean data in table self.values with a safer way

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: t/node/route-delete.t [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+--- no_error_log
+[error]
+--- grep_error_log eval
+qr/\w+ (data by key: 126)/
+--- grep_error_log_out
+insert data by key: 126
+delete data by key: 126
+insert data by key: 126
+delete data by key: 126
+--- timeout: 20
```
Comment (nic-chen) [inline t/node/route-delete.t:93]: has some more test cases  ?

File: apisix/core/config_etcd.lua [hunk 2]
```diff
@@ -314,36 +316,44 @@ local function sync_data(self)
...
     end
 
     -- avoid space waste
     -- todo: need to cover this path, it is important.
     if self.sync_times > 100 then
-        local count = 0
-        for i = 1, #self.values do
-            local val = self.values[i]
-            self.values[i] = nil
+        local values_org = table.clone(self.values)
```
Comment (moonming) [inline apisix/core/config_etcd.lua:339]: What does this suffix mean? `_org`

File: apisix/core/config_etcd.lua [hunk 3]
```diff
@@ -314,36 +316,44 @@ local function sync_data(self)
...
     elseif res.value then
         res.clean_handlers = {}
         insert_tab(self.values, res)
         self.values_hash[key] = #self.values
         res.value.id = key
+        log.info("insert data by key: ", key)
     end
 
     -- avoid space waste
     -- todo: need to cover this path, it is important.
```
Comment (moonming) [inline apisix/core/config_etcd.lua:337]: Is it still a TODO?

File: t/node/route-delete.t (hunk 1 above)
Comment (moonming) [reply t/node/route-delete.t:93]: not cover https://github.com/apache/apisix/pull/2168/files#diff-f2239cc9d5659a13ba363ebc30959916R319

File: apisix/core/config_etcd.lua [hunk 4]
```diff
@@ -314,36 +316,44 @@ local function sync_data(self)
...
     if self.sync_times > 100 then
-        local count = 0
-        for i = 1, #self.values do
-            local val = self.values[i]
-            self.values[i] = nil
+        local values_org = table.clone(self.values)
+        table.clear(self.values)
+
+        for i = 1, #values_org do
```
Comment (moonming) [inline apisix/core/config_etcd.lua:342]: Why do we need to delete nil elements in the array? We can make judgments in this for loop：https://github.com/apache/apisix/pull/2168/files#diff-f2239cc9d5659a13ba363ebc30959916R352

File: apisix/core/config_etcd.lua (hunk 3 above)
Comment (membphis) [reply apisix/core/config_etcd.lua:337]: we can remove this comment now. we have added test cases about it.

File: apisix/core/config_etcd.lua (hunk 4 above)
Comment (membphis) [reply apisix/core/config_etcd.lua:341]: We don't want the "value" table to grow larger and larger, which will cause memory overflow.

Therefore, we need to delete the useless items in the "value" to keep it at an acceptable size.

File: apisix/core/config_etcd.lua (hunk 2 above)
Comment (membphis) [reply apisix/core/config_etcd.lua:338]: `original`

File: t/node/route-delete.t (hunk 1 above)
Comment (membphis) [reply t/node/route-delete.t:109]: fixed

File: apisix/core/config_etcd.lua (hunk 2 above)
Comment (moonming) [reply apisix/core/config_etcd.lua:338]: so `values_original`

Comment (moonming) [reply apisix/core/config_etcd.lua:338]: `org` means organization

**Review by nic-chen — APPROVED** [review]

**Review by moonming — APPROVED** [review]

**Pull request merged**